### ISSUE TITLE
Support adding definitions to the base before starting the scenario (e.g. to implement a DSL)

### ISSUE THREAD

**Issue by kostmo** [swarm-game/swarm#838]
These would be placed in the `program` field of a player's robot in a Scenario.

**Comment by byorgey** [comment]
To clarify, this is already supported in theory, but in practice it seems like attempting this leads to `Fatal error: Reference to unknown memory cell 0`, aka The Bug That Would Not Die: #299 #333 .

**Comment by byorgey** [comment]
Fixed the bug, but now I realized there's another issue: `solution`s cannot use commands defined in the base `program` since solutions are parsed/typechecked in an empty context.  Considering what a good solution to that would be.  Options include (1) parse any `base` `program` first and make its context available when parsing/processing `solution`; (2) make a separate top-level scenario field called `definitions` which will consist of an arbitrary Swarm term (typically consisting of definitions).  This would be similar to specifying a `program` for the base robot but maybe simpler to deal with/reason about if it is a distinguished top-level thing.

**Comment by xsebek** [comment]
@byorgey Is there a reason we can not just concatenate program and solution?

All normal bugs and limitations of swarm terms would of course apply, but I don’t see much benefit to solving them here separately. 🤷

**Comment by byorgey** [comment]
You mean when checking the `solution`, we first just prepend any base `program` to the beginning of it (with a semicolon in between)?  Yes, that would probably work, and is simpler (though perhaps more "hackish") than what I had in mind.

**Comment by xsebek** [comment]
@byorgey Or we could do it after parsing and "concatenate" using the `TBind`. 🙂

**Comment by kostmo** [comment]
This issue may overlap with #778.

**Comment by byorgey** [comment]
This is also related, perhaps, to #495 .  Perhaps we could think in terms of scenarios being able to provide modules that are available for robots to `import` using some special syntax.  i.e. instead of only being able to import things from local disk or a URL, there would be a third option/syntax to import a "scenario-local" module.
